cmd/otel-allocator/allocation/consistent_hashing.go

```diff
@@ -0,0 +1,194 @@
+
+	// targetItems is a map from a target item's hash to the target items allocated state
+	targetItems map[string]*TargetItem
+
+	log logr.Logger
+}
+
+func newConsistentHashingAllocator(log logr.Logger) Allocator {
+	config := consistent.Config{
+		PartitionCount:    1061,
+		ReplicationFactor: 5,
+		Load:              1.1,
+		Hasher:            hasher{},
+	}
+	consistentHasher := consistent.New(nil, config)
+	return &consistentHashingAllocator{
+		consistentHasher: consistentHasher,
+		collectors:       make(map[string]*Collector),
+		targetItems:      make(map[string]*TargetItem),
+		log:              log,
+	}
+}
+
+// addTargetToTargetItems assigns a target to the collector based on its hash and adds it to the allocator's targetItems
+// This method is called from within SetTargets and SetCollectors, which acquire the needed lock.
+// This is only called after the collectors are cleared or when a new target has been found in the tempTargetMap
+// INVARIANT: c.collectors must have at least 1 collector set
+func (c *consistentHashingAllocator) addTargetToTargetItems(target *TargetItem) {
+	// Check if this is a reassignment, if so, decrement the previous collector's NumTargets
+	if previousColName, ok := c.collectors[target.CollectorName]; ok {
+		previousColName.NumTargets--
+		TargetsPerCollector.WithLabelValues(previousColName.String(), consistentHashingStrategyName).Set(float64(c.collectors[previousColName.String()].NumTargets))
+	}
+	colOwner := c.consistentHasher.LocateKey([]byte(target.Hash()))
+	targetItem := &TargetItem{
+		JobName:       target.JobName,
+		Link:          LinkJSON{Link: fmt.Sprintf("/jobs/%s/targets", url.QueryEscape(target.JobName))},
+		TargetURL:     target.TargetURL,
+		Label:         target.Label,
+		CollectorName: colOwner.String(),
+	}
+	c.targetItems[targetItem.Hash()] = targetItem
+	c.collectors[colOwner.String()].NumTargets++
+	TargetsPerCollector.WithLabelValues(colOwner.String(), consistentHashingStrategyName).Set(float64(c.collectors[colOwner.String()].NumTargets))
+}
+
+// handleTargets receives the new and removed targets and reconciles the current state.
+// Any removals are removed from the allocator's targetItems and unassigned from the corresponding collector
+// Any net-new additions are assigned to the next available collector
+func (c *consistentHashingAllocator) handleTargets(diff diff.Changes[*TargetItem]) {
+	// Check for removals
+	for k, target := range c.targetItems {
+		// if the current target is in the removals list
+		if _, ok := diff.Removals()[k]; ok {
+			col := c.collectors[target.CollectorName]
+			col.NumTargets--
+			delete(c.targetItems, k)
+			TargetsPerCollector.WithLabelValues(target.CollectorName, consistentHashingStrategyName).Set(float64(col.NumTargets))
+		}
+	}
+
+	// Check for additions
+	for k, target := range diff.Additions() {
+		// Do nothing if the item is already there
+		if _, ok := c.targetItems[k]; ok {
+			continue
+		} else {
```

Comment:
Is this necessary?  `addTargetToTargetItems()` has provision for when the new target has an existing collector attached to it, which I think would only happen when the target existed previously as well.

Reply:
I believe this is still necessary. the provision in `addTargetToTargetItems()` is for when we reassign a collector after a new collector is added which shifts the amount of members in the hashring. In this case, we are iterating through the newly added targets and seeing if the target has already been added. Because we're using consistent hashing, the target would get unassigned and then reassigned to the same collector which seems like unnecessary work.